xmtp_mls/src/groups/device_sync.rs

```diff
@@ -399,540 +358,514 @@ where
+            let mut add_futs = vec![];
+            for group in chunk {
+                add_futs.push(group.add_missing_installations(&provider));
+            }
+            let results = join_all(add_futs).await;
+            for result in results {
+                if let Err(err) = result {
+                    tracing::warn!("Unable to add new installation to group. {err:?}");
                 }
             }
-            None => Err(DeviceSyncError::Conversion),
         }
+
+        Ok(())
     }
 }
 
-pub(super) fn new_request_id() -> String {
-    xmtp_common::rand_string::<ENC_KEY_SIZE>()
+fn default_backup_options() -> BackupOptions {
+    BackupOptions {
+        elements: vec![
+            BackupElementSelection::Messages as i32,
+            BackupElementSelection::Consent as i32,
+        ],
+        ..Default::default()
+    }
 }
 
-pub(super) fn generate_nonce() -> [u8; NONCE_SIZE] {
-    xmtp_common::rand_array::<NONCE_SIZE>()
+// These are the messages that get sent out to the sync group
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+#[repr(i32)]
+pub enum DeviceSyncContent {
+    Request(DeviceSyncRequestProto) = 0,
+    Payload(DeviceSyncReplyProto) = 1,
+    Acknowledge(AcknowledgeKind) = 2,
+    PreferenceUpdates(Vec<UserPreferenceUpdate>) = 3,
 }
-
-pub(super) fn new_pin() -> String {
-    let mut rng = crypto_utils::rng();
-    let pin: u32 = rng.gen_range(0..10000);
-    format!("{:04}", pin)
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub enum AcknowledgeKind {
+    SyncGroupPresence,
+    Request { request_id: String },
 }
 
-fn encrypt_syncables(
-    syncables: &[Vec<Syncable>],
-) -> Result<(Vec<u8>, DeviceSyncKeyType), DeviceSyncError> {
-    let enc_key = DeviceSyncKeyType::new_aes_256_gcm_key();
-    encrypt_syncables_with_key(syncables, enc_key)
+pub trait ZipContent<A, B> {
+    fn iter_with_content(self) -> impl Iterator<Item = (A, B)>;
 }
 
-fn encrypt_syncables_with_key(
-    syncables: &[Vec<Syncable>],
-    enc_key: DeviceSyncKeyType,
-) -> Result<(Vec<u8>, DeviceSyncKeyType), DeviceSyncError> {
-    let syncables: Vec<&Syncable> = syncables.iter().flat_map(|s| s.iter()).collect();
-    let payload = serde_json::to_vec(&syncables)?;
-
-    let enc_key_bytes = enc_key.as_bytes();
-    let mut result = generate_nonce().to_vec();
-
-    // create a cipher instance
-    let cipher = Aes256Gcm::new(GenericArray::from_slice(enc_key_bytes));
-    let nonce_array = GenericArray::from_slice(&result);
-
-    // encrypt the payload and append to the result
-    result.append(&mut cipher.encrypt(nonce_array, &*payload)?);
-
-    Ok((result, enc_key))
+impl ZipContent<StoredGroupMessage, DeviceSyncContent> for Vec<StoredGroupMessage> {
+    fn iter_with_content(self) -> impl Iterator<Item = (StoredGroupMessage, DeviceSyncContent)> {
+        self.into_iter().filter_map(|msg| {
+            let content: DeviceSyncContent =
+                serde_json::from_slice(&msg.decrypted_message_bytes).ok()?;
```

Comment:
using `.ok()` here could create a situation where we're de-serializing something incorrectly but ignoring the item in an iterator, resulting in something looking like we're missing a message, which could be difficult do debug. We should prefer using inspect() and at the very least trace/debug logging a warning that something failed to deserialize, rather then ignore the option

Reply:
Yeah, this should be logged.